homeassistant/components/media_player/epson.py

```diff
@@ -0,0 +1,245 @@
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_PORT, default=80): cv.port,
+    vol.Optional(CONF_SSL, default=False): cv.boolean
+})
+
+ATTR_CMODE = 'cmode'
+SUPPORT_CMODE = 33001
+
+SUPPORT_EPSON = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE |\
+            SUPPORT_CMODE | SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_STEP | \
+            SUPPORT_NEXT_TRACK | SUPPORT_PREVIOUS_TRACK
+
+EPSON_SCHEMA = MEDIA_PLAYER_SCHEMA.extend({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Optional(ATTR_CMODE): cv.string
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(
+        hass,
+        config,
+        async_add_devices,
+        discovery_info=None):
+    """Set up the Epson media player platform."""
+    if DATA_EPSON not in hass.data:
+        hass.data[DATA_EPSON] = []
+    name = config.get(CONF_NAME)
+    host = config.get(CONF_HOST)
+
+    try:
+        epson = EpsonProjector(
+            hass, name, host,
+            config.get(CONF_PORT), config.get(CONF_SSL))
+        await epson.update()
+    except (aiohttp.client_exceptions.ClientConnectorError,
+            asyncio.TimeoutError):
+        raise PlatformNotReady
+    if epson:
+        hass.data[DATA_EPSON].append(epson)
+        async_add_devices([epson], update_before_add=True)
+
+    async def async_service_handler(service):
+        """Handle for services."""
+        from epson_projector.const import (CMODE_LIST_SET)
+        entity_ids = service.data.get(ATTR_ENTITY_ID)
+        if entity_ids:
+            devices = [device for device in hass.data[DATA_EPSON]
+                       if device.entity_id in entity_ids]
+        else:
+            devices = hass.data[DATA_EPSON]
+        for device in devices:
+            if service.service == ATTR_CMODE:
+                cmode = service.data.get(ATTR_CMODE).lower()
+                if cmode in CMODE_LIST_SET:
+                    await device.select_cmode(cmode)
+            await device.update()
+    hass.services.async_register(
+        DOMAIN, ATTR_CMODE, async_service_handler,
+        schema=EPSON_SCHEMA)
+    return True
+
+
+class EpsonProjector(MediaPlayerDevice):
+    """Representation of Epson Projector Device."""
+
+    def __init__(self, hass, name, host, port, encryption):
+        """Initialize entity to control Epson projector."""
+        self._hass = hass
+        self._name = name
+        import epson_projector as epson
+        from epson_projector.const import DEFAULT_SOURCES
+        self._projector = epson.Projector(
+            host,
+            websession=self._websession(False),
+            port=port)
+
+        self._cmode = None
+        self._source_list = list(DEFAULT_SOURCES.values())
+        self._source = None
+        self._volume = None
+
+        self._state = None
+
+    def _websession(self, verify_ssl):
+        """Return a websession."""
+        return async_get_clientsession(self._hass, verify_ssl)
```

Comment:
Move this to platform setup and pass the session to the entity instance instead.